keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,548 @@
+
+1. When the leader received member attribute update from raft, it will decide the values for the cluster level features based on the proposed feature values of all members.
+
+1. The leader sends the final feature values in `ClusterFeatureGateSetRequest` through raft. 
+
+1. Each member apply the updates to their cluster feature gate, and save the results in the `cluster` bucket in the backend.
+
+![cluster feature gate consensus algorithm](./cluster_feature_gate.png "cluster feature gate consensus algorithm")
+
+#### New Raft Proto Changes
+
+```proto
+message Feature {
+  string name = 1;
+  bool enabled = 2;
+}
+
+message Attributes {
+  option (versionpb.etcd_version_msg) = "3.5";
+
+  string name = 1;
+  repeated string client_urls = 2;
+  // the values of all cluster level feature gates set by the configuration of the member server.
+  repeated Feature proposed_feature_gates = 3;
+}
+
+message ClusterFeatureGateSetRequest {
+  option (versionpb.etcd_version_msg) = "3.5";
+
+  // the values of all cluster level feature gates for the whole cluster.
+  repeated Feature feature_gates = 1;
+}
+```
+
+#### New Backend Schema Changes
+We will save the values of cluster feature_gates in the `Cluster` bucket in the backend.
+
+#### Start a New Cluster
+
+Fig. a above shows how the consensus of feature gates is reached for a new cluster:
+1. When the cluster starts, before `ClusterVersion` is set, the cluster feature gate would be `nil`, and all features would be considered disabled. 
+1. Before the leader send an update message of the `ClusterVersion`, the leader will send a `ClusterFeatureGateSetRequest` message with empty `feature_gates` to reset any existing cluster feature gate. This is to make sure the feature gate is always compatible with the `ClusterVersion`.
+1. Whenever the `ClusterVersionSet` is applied in a server, the member will propose the cluster feature gate values based on its own flags and the `ClusterVersion`.
+1. Whenever the leader receives the `member.Attributes.proposed_feature_gates` from a member, it would try to do `UpdateClusterFeatureGate` by reconciling the `proposed_feature_gates` from all the members: 
+   * if the `proposed_feature_gates` from any member is missing, the cluster feature gate will be empty, i.e. all features will be disabled.
+     * [NOTE] the leader would wait for all the members to send in their `proposed_feature_gates` instead of just initialize the feature gate to the default values at the `ClusterVersion`, because different members might have different patch versions, and may not have the same default values. We do not want the feature values to flip-flop when the leader changes.
```

Comment:
I understand that running adding member with different patch version with different defaults might cause feature to flip, but I don't see how leader defaults could cause flip flops.